File: leet_LIST_0092_reverse_ll_ii___inplace_rev.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* reverseBetween(struct ListNode* head, int left, int right)
{
    //struct ListNode     *right_list; // no use eventually
    struct ListNode     *left_list;
    struct ListNode     *mid_list;
    struct ListNode     *prev;
    struct ListNode     *next;
    struct ListNode     *self;
    int                 i;

    if (head == NULL)
        return (head);
    i = 0;
    self = head;
    left_list = NULL;
    while (++i < left) /* 1st traversal */
    {
        left_list = self;
        self = self->next;
    }
    mid_list = self;
    while (self && i < right + 1)
    {
        next = self->next;
        self->next = prev;
        prev = self;
        self = next;
        ++i;
    }
    if (left_list)
        left_list->next = prev; /* if 1st traversal did happen */
    else
        head = prev; /* 1st traversal did not happen, ie. only mid section is reversed */
    mid_list->next = self;
    return (head);
}

File: leet_LIST_0092_reverse_ll_ii___inplace_rev.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function(head, left, right) {
    if (!head)
        return head
    let left_end = null
    let self = head
    let i = 0
    while (++i < left) {
        left_end = self
        self = self.next
    }
    let mid_head = null
    let mid_end = self // ie. "prev" in reverse_linked_list
    let next = null
    while (self && i++ < right + 1) {
        next = self.next
        self.next = mid_head
        mid_head = self
        self = next
    }
    if (left_end) {
        left_end.next = mid_head
    } else {
        head = mid_head
    }
    mid_end.next = self
    return head
};
